rdr_service/dao/questionnaire_response_dao.py

```diff
@@ -808,6 +823,49 @@ def _update_participant_summary(
                     session, participant.participantId, gender_code_ids
                 )
 
+    def create_consent_responses(self, questionnaire_response: QuestionnaireResponse, session: Session):
+        """
+        Analyzes the current ConsentResponses for a participant, and the response that was just received
+        to determine if the new response is a new consent for the participant.
+        """
+        if len(self.consents_provided) == 0:
+            # If the new response doesn't give any consent at all, then there's no reason for a check
+            return
+
+        # Load previously received consent authored dates for the participant
+        previous_consent_dates = ConsentDao.get_consent_authored_times_for_participant(
+            session=session,
+            participant_id=questionnaire_response.participantId
+        )
+
+        # Check authored dates to see if it's a new consent response,
+        # or if it's potentially just a replay of a previous questionnaire response
+        for consent_type in self.consents_provided:
+            is_new_consent = False
+            previous_authored_times = previous_consent_dates.get(consent_type)
+            if not previous_authored_times:  # Brand new consent response
+                is_new_consent = True
+            else:
+                for previous_consent_authored_time in previous_authored_times:
+                    if not self._authored_times_match(
+                        new_authored_time=questionnaire_response.authored,
+                        current_authored_item=previous_consent_authored_time
+                    ):
+                        is_new_consent = True
+                        break
+
+            if is_new_consent:
+                session.add(ConsentResponse(response=questionnaire_response, type=consent_type))
+
+    @classmethod
+    def _authored_times_match(cls, new_authored_time: datetime, current_authored_item: datetime):
+        if new_authored_time.tzinfo is None:
+            new_authored_time = new_authored_time.replace(tzinfo=pytz.utc)
+        if current_authored_item.tzinfo is None:
+            current_authored_item = current_authored_item.replace(tzinfo=pytz.utc)
+        difference_in_seconds = abs((new_authored_time - current_authored_item).total_seconds())
+        return difference_in_seconds < 300  # Allowing 5 minutes of difference between authored dates
```

Comment:
So I understand, this 5 minute window would allow for potentially multiple QuestionnaireResponse records that may not be/are not yet flagged as duplicates;  but for consent validation purposes, it's assumed they relate to the same consent PDF file if the authored times are so close?

Reply:
I'm wondering more about PTSC's potential ability to send extraneous QuestionnaireResponse payloads (symptomatic of a process error) within a 5-minute window.   (Not necessarily duplicates and/or potentially referring to multiple PDFs that may have been improperly generated).

Would such a process error be easily detectable on our end if we only record one of them as a new consent response?  Is it worth logging a warning if we do detect "matching" authored times within the 5-minute window?

Reply:
If I'm understanding correctly, you're suggesting that this be used to also detect if a QuestionnaireResponse is being received multiple times when it shouldn't be. I think that's a good idea to implement, but that will take some careful thought and design to implement well, and that isn't the purpose of the changes I'm working on here.